_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+import React from "react";
+
+const Auth0ProviderWithNavigate = (props: { children: React.ReactNode }) => {
+  const router = useRouter();
+
+  const domain = process.env.NEXT_PUBLIC_AUTH0_DOMAIN || "";
+  const clientId = process.env.NEXT_PUBLIC_AUTH0_CLIENT_ID || "";
+  const redirectUri = process.env.NEXT_PUBLIC_AUTH0_CALLBACK_URL || "";
+  const audience = process.env.NEXT_PUBLIC_AUTH0_AUDIENCE || "";
+
+  const onRedirectCallback = (appState?: AppState) => {
+    router.push(appState?.returnTo || window.location.pathname);
+  };
+
+  if (!(domain && clientId && redirectUri)) {
+    return null;
+  }
+
+  return (
+    <Auth0Provider
+      domain={domain}
+      clientId={clientId}
+      authorizationParams={{
+        audience: audience,
+        redirect_uri: redirectUri,
+      }}
+      useRefreshTokens={true}
+      onRedirectCallback={onRedirectCallback}
+    >
+      <>{props.children}</>
+    </Auth0Provider>
+  );
+};
+
+export default Auth0ProviderWithNavigate;
+```
+{% endraw %}
+
+The component `Auth0ProviderWithNavigate` wraps the children component with `Auth0Provider`, the provider of the Auth0 context, remembering the requested URL for redirection after login.
+Use the component in the `WideLayout` component:
+
+{% raw %}
+__WideLayout.tsx__
+```tsx
+const WideLayout = (props: { children: React.ReactNode }) => {
+  return (
+    <ThemeProvider theme={theme}>
+      <Auth0ProviderWithNavigate>
+        <Container maxWidth="lg" sx={{ mt: 4 }}>
+          {props.children}
+        </Container>
+      </Auth0ProviderWithNavigate>
+    </ThemeProvider>
+  );
+};
+```
+{% endraw %}
+
+Add the file `src/components/authentication/AuthenticationGuard.tsx` with the following content:
+
+__AuthenticationGuard.tsx__
+```tsx
+'use client'
+
+import { useAuth0 } from "@auth0/auth0-react";
+import { useEffect } from "react";
+import Loader from "../loader/Loader";
+
+const AuthenticationGuard = (props: { children: React.ReactNode }) => {
+  const { isLoading, isAuthenticated, error, loginWithRedirect } = useAuth0();
+
+  useEffect(() => {
+    if (!isAuthenticated && !isLoading) {
+      loginWithRedirect({
+        appState: { returnTo: window.location.href },
+      });
+    }
+  }, [isAuthenticated, isLoading, loginWithRedirect]);
+
+  if (isLoading) {
+    return <Loader />;
+  }
+  if (error) {
+    return <div>Oops... {error.message}</div>;
+  }
+  return <>{isAuthenticated && props.children}</>;
+};
+
+export default AuthenticationGuard;
+```
+
+The `AuthenticationGuard` component will be used to protect pages that require authentication, redirecting to the Auth0 universal login. Protect the index page by wrapping its content in the `AuthenticationGuard` component:
+
+{% raw %}
+__app/page.tsx__
+```tsx
+import AuthenticationGuard from "@/components/authentication/AuthenticationGuard";
+import HomePage from "./HomePage";
+
+const Page = () => {
+  return (
+    <AuthenticationGuard>
+      <HomePage></HomePage>
```

Comment:
A recent version of React made it possible to do `<HomePage/>`

Reply:
Updated